File: server/snowglobe/k8s.py
"""Live cluster state via kubectl. No Kubernetes client library needed:
kubectl handles auth (exec plugins, EKS/GKE/AKS helpers) for us."""
import json
import math
import shutil
import subprocess

COLOR_CYCLE = ["#3b7dd8", "#1d9e75", "#d85a30", "#7a5cc4", "#c2417e", "#2a8fa3"]


def kubectl_available():
    return shutil.which("kubectl") is not None


def _run(args, timeout=15):
    try:
        out = subprocess.run(
            ["kubectl"] + args, capture_output=True, text=True, timeout=timeout
        )
        if out.returncode != 0:
            return None
        return out.stdout
    except (subprocess.TimeoutExpired, OSError):
        return None


def _pod_phase(pod):
    statuses = (pod.get("status", {}).get("containerStatuses") or [])
    for cs in statuses:
        waiting = (cs.get("state") or {}).get("waiting") or {}
        if waiting.get("reason") in ("CrashLoopBackOff", "ImagePullBackOff", "ErrImagePull"):
            return "crash"
    if pod.get("status", {}).get("phase") == "Pending":
        return "pending"
    return "ok"


def _pod_restarts(pod):
    return sum(cs.get("restartCount", 0) for cs in (pod.get("status", {}).get("containerStatuses") or []))


def _cpu_millicores(value):
    if not value:
        return 0
    if value.endswith("n"):
        return int(value[:-1]) / 1_000_000
    if value.endswith("m"):
        return int(value[:-1])
    return float(value) * 1000


def fetch_cluster(namespace):
    ns_args = ["-n", namespace] if namespace else ["--all-namespaces"]

    deploys_raw = _run(["get", "deployments", "-o", "json"] + ns_args)
    pods_raw = _run(["get", "pods", "-o", "json"] + ns_args)
    if deploys_raw is None or pods_raw is None:
        return None

    deploys = json.loads(deploys_raw).get("items", [])
    pods = json.loads(pods_raw).get("items", [])

    top = {}
    top_raw = _run(["top", "pods", "--no-headers"] + ns_args, timeout=10)
    if top_raw:
        for line in top_raw.strip().splitlines():
            parts = line.split()
            if namespace and len(parts) >= 2:
                top[parts[0]] = _cpu_millicores(parts[1])
            elif len(parts) >= 3:
                top[parts[1]] = _cpu_millicores(parts[2])

    services = []
    n = max(1, len(deploys))
    for i, dep in enumerate(deploys[:12]):
        dep_name = dep["metadata"]["name"]
        desired = dep.get("spec", {}).get("replicas", 1) or 1

        dep_pods = []
        for pod in pods:
            if not pod["metadata"]["name"].startswith(dep_name + "-"):
                continue
            name = pod["metadata"]["name"]
            limit = 0
            for c in pod.get("spec", {}).get("containers", []):
                lim = (c.get("resources", {}).get("limits") or {}).get("cpu")
                if lim:
                    limit += _cpu_millicores(lim)
            used = top.get(name, 0)
            util = used / limit if limit else min(1.0, used / 500)
            dep_pods.append({
                "id": name,
                "phase": _pod_phase(pod),
                "util": round(util, 3),
                "targetUtil": round(util, 3),
                "restarts": _pod_restarts(pod),
                "ticksInPhase": 0,
            })

        angle = -math.pi + (2 * math.pi * i) / n
        services.append({
            "name": dep_name,
            "color": COLOR_CYCLE[i % len(COLOR_CYCLE)],
            "angle": round(angle, 3),
            "dist": 9.5,
            "desired": desired,
            "pods": dep_pods or [{
                "id": f"{dep_name}-none", "phase": "pending", "util": 0,
                "targetUtil": 0, "restarts": 0, "ticksInPhase": 0,
            }],
            "rps": 0,
            "upstream": None,
        })

    context = (_run(["config", "current-context"], timeout=5) or "cluster").strip()
    return {"name": context, "gatewayName": "kubernetes", "services": services}

File: cli/src/k8s.ts
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'
import { ClusterState, FetchReason, FetchResult, PodInfo, ServiceInfo } from './types.js'

const exec = promisify(execFile)

const COLOR_CYCLE = ['#3b7dd8', '#1d9e75', '#d85a30', '#7a5cc4', '#c2417e', '#2a8fa3']

const SYSTEM_NS = new Set([
  'kube-system', 'kube-public', 'kube-node-lease',
  'gmp-system', 'gmp-public', 'gke-managed-system', 'gke-managed-cim',
  'gke-managed-volumepopulator', 'gke-gmp-system',
])

const MAX_WORKLOADS = 24
const MAX_PODS_PER_WORKLOAD = 40

interface RunResult {
  ok: boolean
  stdout: string
  stderr: string
}

async function run(args: string[], timeoutMs = 15000): Promise<RunResult> {
  try {
    const { stdout } = await exec('kubectl', args, {
      timeout: timeoutMs,
      maxBuffer: 128 * 1024 * 1024,
    })
    return { ok: true, stdout, stderr: '' }
  } catch (e: any) {
    return { ok: false, stdout: e?.stdout ?? '', stderr: e?.stderr ?? String(e?.message ?? e) }
  }
}

function classify(stderr: string): FetchReason {
  const s = stderr.toLowerCase()
  if (s.includes('executable file not found') || s.includes('enoent') || s.includes('command not found')) {
    return 'no-kubectl'
  }
  if (s.includes('forbidden') || s.includes('cannot list') || s.includes('is not allowed')) {
    return 'forbidden'
  }
  return 'unreachable'
}

export async function kubectlAvailable(): Promise<boolean> {
  const r = await run(['version', '--client', '-o', 'json'], 5000)
  return r.ok || classify(r.stderr) !== 'no-kubectl'
}

function cpuMillicores(value: string | undefined): number {
  if (!value) return 0
  if (value.endsWith('n')) return parseInt(value) / 1_000_000
  if (value.endsWith('u')) return parseInt(value) / 1000
  if (value.endsWith('m')) return parseInt(value)
  return parseFloat(value) * 1000
}

function podPhase(pod: any): 'ok' | 'crash' | 'pending' {
  for (const cs of pod?.status?.containerStatuses ?? []) {
    const reason = cs?.state?.waiting?.reason
    if (reason === 'CrashLoopBackOff' || reason === 'ImagePullBackOff' || reason === 'ErrImagePull') {
      return 'crash'
    }
  }
  const phase = pod?.status?.phase
  if (phase === 'Pending') return 'pending'
  if (phase === 'Failed') return 'crash'
  return 'ok'
}

function podRestarts(pod: any): number {
  let total = 0
  for (const cs of pod?.status?.containerStatuses ?? []) total += cs?.restartCount ?? 0
  return total
}

function selectorMatches(selector: Record<string, string>, labels: Record<string, string>): boolean {
  const keys = Object.keys(selector)
  if (keys.length === 0) return false
  return keys.every((k) => labels?.[k] === selector[k])
}

// crash/pending first so problems stay visible even when a workload is capped
const PHASE_RANK: Record<string, number> = { crash: 0, pending: 1, ok: 2 }

async function getItems(resource: string, nsArgs: string[]): Promise<{ items: any[]; err?: RunResult }> {
  const r = await run(['get', resource, '-o', 'json', ...nsArgs])
  if (!r.ok) return { items: [], err: r }
  try {
    return { items: JSON.parse(r.stdout).items ?? [] }
  } catch {
    return { items: [] }
  }
}

export async function fetchCluster(namespace?: string): Promise<FetchResult> {
  const nsArgs = namespace ? ['-n', namespace] : ['--all-namespaces']

  const [deploys, statefulsets, daemonsets, podsRes] = await Promise.all([
    getItems('deployments', nsArgs),
    getItems('statefulsets', nsArgs),
    getItems('daemonsets', nsArgs),
    getItems('pods', nsArgs),
  ])

  // Pods are the ground truth; if we can't read them, that's the real failure.
  if (podsRes.err) {
    const reason = classify(podsRes.err.stderr)
    return { ok: false, reason, message: podsRes.err.stderr.trim().split('\n').pop() || 'kubectl could not reach a cluster' }
  }
  // If every workload query also failed with a permission error, surface that.
  if (deploys.err && statefulsets.err && daemonsets.err) {
    const reason = classify(deploys.err.stderr)
    if (reason === 'forbidden' || reason === 'no-kubectl') {
      return { ok: false, reason, message: deploys.err.stderr.trim().split('\n').pop() || 'kubectl could not reach a cluster' }
    }
  }

  const keep = (item: any) => namespace || !SYSTEM_NS.has(item?.metadata?.namespace)
  const workloads = [
    ...deploys.items.filter(keep).map((w) => ({ w, kind: 'Deployment' })),
    ...statefulsets.items.filter(keep).map((w) => ({ w, kind: 'StatefulSet' })),
    ...daemonsets.items.filter(keep).map((w) => ({ w, kind: 'DaemonSet' })),
  ]

  const pods = podsRes.items

  // metrics-server: if `kubectl top` fails, CPU is simply unavailable.
  const top = new Map<string, number>()
  const topRes = await run(['top', 'pods', '--no-headers', ...nsArgs], 10000)
  const metricsAvailable = topRes.ok
  if (topRes.ok) {
    for (const line of topRes.stdout.trim().split('\n')) {
      if (!line.trim()) continue
      const parts = line.split(/\s+/)
      if (namespace && parts.length >= 2) top.set(parts[0], cpuMillicores(parts[1]))
      else if (parts.length >= 3) top.set(parts[1], cpuMillicores(parts[2]))
    }
  }

  const workloadTotal = workloads.length
  const shown = workloads.slice(0, MAX_WORKLOADS)
  const n = Math.max(1, shown.length)
  const ringDist = Math.max(9.5, (n * 7.5) / (2 * Math.PI))

  const services: ServiceInfo[] = shown.map(({ w, kind }, i) => {
    const selector: Record<string, string> = w.spec?.selector?.matchLabels ?? {}
    let matched = pods.filter((p: any) => selectorMatches(selector, p.metadata?.labels ?? {}))
    matched.sort((a: any, b: any) => PHASE_RANK[podPhase(a)] - PHASE_RANK[podPhase(b)])

    let desired =
      kind === 'DaemonSet'
        ? w.status?.desiredNumberScheduled ?? (matched.length || 1)
        : w.spec?.replicas ?? 1

    const capped = matched.length > MAX_PODS_PER_WORKLOAD
    if (capped) {
      matched = matched.slice(0, MAX_PODS_PER_WORKLOAD)
      desired = Math.min(desired, MAX_PODS_PER_WORKLOAD)
    }

    const pod = (p: any): PodInfo => {
      let limit = 0
      for (const c of p.spec?.containers ?? []) limit += cpuMillicores(c.resources?.limits?.cpu)
      const used = top.get(p.metadata.name) ?? 0
      const util = metricsAvailable && limit ? Math.min(1, used / limit) : 0
      return {
        id: p.metadata.name,
        phase: podPhase(p),
        util: Math.round(util * 1000) / 1000,
        targetUtil: Math.round(util * 1000) / 1000,
        restarts: podRestarts(p),
        ticksInPhase: 0,
      }
    }

    return {
      name: w.metadata.name,
      kind,
      color: COLOR_CYCLE[i % COLOR_CYCLE.length],
      angle: Math.round((-Math.PI + (2 * Math.PI * i) / n) * 1000) / 1000,
      dist: Math.round(ringDist * 100) / 100,
      desired,
      pods: matched.length
        ? matched.map(pod)
        : [{ id: `${w.metadata.name}-none`, phase: 'pending', util: 0, targetUtil: 0, restarts: 0, ticksInPhase: 0 }],
      rps: 0,
      upstream: null,
    }
  })

  if (services.length === 0) {
    return {
      ok: false,
      reason: 'empty',
      message: namespace
        ? `no workloads found in namespace "${namespace}"`
        : 'no workloads found (only system namespaces?)',
    }
  }

  const ctxRes = await run(['config', 'current-context'], 5000)
  const context = ctxRes.ok ? ctxRes.stdout.trim() : 'cluster'

  const cluster: ClusterState = {
    name: context,
    gatewayName: 'kubernetes',
    services,
    metricsAvailable,
    workloadTotal,
  }
  return { ok: true, cluster }
}
